geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/BinnedArray.scala

```diff
@@ -274,83 +290,52 @@ class BinnedDoubleArray(length: Int, bounds: (jDouble, jDouble)) extends BinnedA
     }
     (bounds._1 + binSize * index, bounds._1 + binSize * (index + 1))
   }
+
+  override def isBelow(value: jDouble): Boolean = value < bounds._1
 }
 
 /**
   * Bins strings. Will set up bins based on the longest-common-prefix of the bounds. Estimates
   * bins by considering inputs to be roughly equivalent to base36 longs.
   *
   * @param length number of bins
-  * @param bounds upper and lower bounds for the input values
+  * @param rawBounds upper and lower bounds for the input values
   */
-class BinnedStringArray(length: Int, bounds: (String, String)) extends BinnedArray[String](length, bounds) {
-
-  private val (start, end) = {
-    val lower = normalize(bounds._1)
-    val upper = normalize(bounds._2)
-    if (lower.length == upper.length) {
-      (lower, upper)
-    } else if (lower.length < upper.length) {
-      (lower.padTo(upper.length, '0'), upper)
-    } else {
-      (lower, upper.padTo(lower.length, 'z'))
+class BinnedStringArray(length: Int, rawBounds: (String, String))
+    extends WholeNumberBinnedArray[String](length, BinnedStringArray.normalizeBounds(rawBounds)) {
+
+  private lazy val (start, end): (String, String) = bounds
+  private lazy val prefixLength = start.zip(end).indexWhere { case (l, r) => l != r }
+  private lazy val prefix = start.substring(0, prefixLength)
+
+  override protected def convertToLong(value: String): Long = {
+    val normalized = BinnedStringArray.normalize(value)
+    if (normalized < start) { 0L } else if (normalized > end) { Long.MaxValue } else {
+      // note: 12 is the most base-36 numbers we can fit in Long.MaxValue
+      val sigDigits = normalized.substring(prefixLength).padTo(12, '0').substring(0, 12)
```

Comment:
yeah, padTo appends on the right side. But in this case, that's what we want, as otherwise almost every string that's not 12 digits will end up being on the very low end of the range and probably sort into the first bucket.